### PR TITLE
wgengine/magicsock: fix data race on endpoint.discoKey

### PR DESCRIPTION
**Pull request by josharian**
endpoint.discoKey is protected by endpoint.mu.
endpoint.sendDiscoMessage was reading it without holding the lock.
This showed up in a CI failure and is readily reproducible locally.

The fix is in two parts.

First, for Conn.enqueueCallMeMaybe, eliminate the one-line helper method endpoint.sendDiscoMessage; call Conn.sendDiscoMessage directly.
This makes it more natural to read endpoint.discoKey in a context
in which endpoint.mu is already held.

Second, for endpoint.sendDiscoPing, explicitly pass the disco key
as an argument. Again, this makes it easier to read endpoint.discoKey
in a context in which endpoint.mu is already held.

Signed-off-by: Josh Bleecher Snyder <josh@tailscale.com>

### CODE REVIEW

**Comment by josharian** [comment]
@danderson @bradfitz do you think we should try to squeeze this into 1.18? It's not obvious to me how big a risk this poses in practice, and how that risk compares to the risk of pulling in this change.

**Comment by bradfitz** [comment]
> @danderson @bradfitz do you think we should try to squeeze this into 1.18? It's not obvious to me how big a risk this poses in practice, and how that risk compares to the risk of pulling in this change.

Yes, we should. It's a regression from this cycle and data races aren't okay; I don't want to ever debug an issue and wonder if it's race-related.

File: wgengine/magicsock/magicsock.go [hunk 1]
```diff
@@ -3544,13 +3544,16 @@ func (de *endpoint) removeSentPingLocked(txid stun.TxID, sp sentPing) {
 	delete(de.sentPing, txid)
 }
 
-// sendDiscoPing sends a ping with the provided txid to ep.
+// sendDiscoPing sends a ping with the provided txid to ep using discoKey.
```
Comment (bradfitz) [inline wgengine/magicsock/magicsock.go:3547]: clarify whose discoKey this is. ours vs ep's.

**Review by bradfitz — APPROVED** [review]

File: wgengine/magicsock/magicsock.go (hunk 1 above)
Comment (josharian) [reply wgengine/magicsock/magicsock.go:3547]: Done.

**Pull request merged**